Content.Server/Atmos/Piping/Unary/EntitySystems/GasThermoMachineSystem.cs

```diff
@@ -50,34 +50,46 @@ public override void Initialize()
         private void OnThermoMachineUpdated(EntityUid uid, GasThermoMachineComponent thermoMachine, AtmosDeviceUpdateEvent args)
         {
 
-            if (!(_power.IsPowered(uid))
+            if (!(_power.IsPowered(uid) && TryComp(uid, out ApcPowerReceiverComponent? receiver))
                 || !TryComp(uid, out NodeContainerComponent? nodeContainer)
                 || !_nodeContainer.TryGetNode(nodeContainer, thermoMachine.InletName, out PipeNode? inlet))
             {
                 return;
             }
 
-            var airHeatCapacity = _atmosphereSystem.GetHeatCapacity(inlet.Air);
-            var combinedHeatCapacity = airHeatCapacity + thermoMachine.HeatCapacity;
+            float sign = Math.Sign(thermoMachine.Cp); // 1 if heater, -1 if freezer
 
-            var startEnergy = inlet.Air.Temperature * airHeatCapacity;
-
-            if (!MathHelper.CloseTo(combinedHeatCapacity, 0, 0.001f))
+            // Implement hysteresis
+            float hystTemp = thermoMachine.TargetTemperature;
+            if (thermoMachine.HysteresisState)
             {
-                var combinedEnergy = thermoMachine.HeatCapacity * thermoMachine.TargetTemperature + airHeatCapacity * inlet.Air.Temperature;
-                inlet.Air.Temperature = combinedEnergy / combinedHeatCapacity;
+                hystTemp += sign*thermoMachine.TemperatureTolerance;
+            }
 
-                thermoMachine.LastEnergyDelta = inlet.Air.Temperature * airHeatCapacity - startEnergy;
+            // Stop if we've reached our target temperature.
+            if (sign*inlet.Air.Temperature > sign*hystTemp) // inequality flips if freezer
+            {
+                thermoMachine.HysteresisState = !thermoMachine.HysteresisState;
+                return;
             }
+
+            // Multiply power in by coefficient of performance, add that heat to gas
+            float dQ = receiver.Load * thermoMachine.Cp * args.dt;
```

Comment:
same

Reply:
This is convention in physics. *Q* is heat. *dQ* is the change in heat. If it was not obvious to the reader, the comment immediately above should make it so.